File: mobile-flutter/lib/src/services/realtime_service.dart
import 'dart:async';
import 'dart:convert';

import 'package:web_socket_channel/web_socket_channel.dart';

import '../core/config.dart';
import '../models/chat_room.dart';
import '../models/user_profile.dart';
import 'chat_api.dart';

class RealtimeService {
  RealtimeService(this._api);

  final ChatApi _api;
  WebSocketChannel? _socket;
  StreamSubscription<dynamic>? _subscription;

  Future<void> connect({
    required UserProfile user,
    required ChatRoom room,
    required void Function(Map<String, dynamic> event) onEvent,
    required void Function() onDisconnected,
  }) async {
    await close();

    final ticket = await _api.createWebSocketTicket();
    final uri = Uri.parse(wsBaseUrl).replace(
      queryParameters: {
        'ticket': ticket,
        'conversation_id': room.conversationId,
      },
    );
    final socket = WebSocketChannel.connect(uri);
    _socket = socket;
    _subscription = socket.stream.listen(
      (rawEvent) {
        final event = jsonDecode(rawEvent.toString());
        if (event is Map<String, dynamic>) {
          onEvent(event);
        }
      },
      onDone: onDisconnected,
      onError: (_) => onDisconnected(),
    );
  }

  void sendActiveConversation(ChatRoom room) {
    _socket?.sink.add(
      jsonEncode({
        'type': 'active_conversation',
        'conversation_id': room.conversationId,
      }),
    );
  }

  void sendMessage({
    required UserProfile user,
    required ChatRoom room,
    required String text,
  }) {
    _socket?.sink.add(
      jsonEncode({
        'sender': user.displayName,
        'sender_id': user.id,
        'recipient_id': room.recipientId,
        'conversation_id': room.conversationId,
        'text': text,
        'attachment_url': '',
        'attachment_name': '',
        'attachment_type': '',
        'attachment_size': 0,
      }),
    );
  }

  Future<void> close() async {
    await _subscription?.cancel();
    await _socket?.sink.close();
    _subscription = null;
    _socket = null;
  }
}

File: mobile-react-native/src/services/realtimeService.js
import { wsBaseUrl } from '../config/runtime'

export function connectRealtime({ profile, room, onEvent, onOpen, onClose, onError }) {
  const url = `${wsBaseUrl}?user_id=${encodeURIComponent(
    profile.id,
  )}&conversation_id=${encodeURIComponent(room.conversationId)}`
  const socket = new WebSocket(url)

  socket.onopen = () => onOpen?.({ url })
  socket.onclose = (event) =>
    onClose?.({
      url,
      code: event?.code,
      reason: event?.reason || '',
      wasClean: Boolean(event?.wasClean),
    })
  socket.onerror = (event) => onError?.({ url, event })
  socket.onmessage = (event) => {
    try {
      const socketEvent = JSON.parse(event.data)
      onEvent(socketEvent, profile.id)
    } catch {
      onError?.({ url, event, reason: 'invalid_message' })
    }
  }

  return socket
}

export function sendActiveConversation(socket, conversationId) {
  socket.send(JSON.stringify({ type: 'active_conversation', conversation_id: conversationId }))
}

export function sendRealtimeMessage(socket, { profile, room, text, attachment }) {
  socket.send(
    JSON.stringify({
      sender: profile.displayName,
      sender_id: profile.id,
      recipient_id: room.recipientId,
      conversation_id: room.conversationId,
      text,
      attachment_url: attachment?.url || '',
      attachment_name: attachment?.name || '',
      attachment_type: attachment?.type || '',
      attachment_size: attachment?.size || 0,
    }),
  )
}

export function closeRealtime(socketRef) {
  if (socketRef.current) {
    socketRef.current.onclose = null
    socketRef.current.onerror = null
    socketRef.current.onmessage = null
    socketRef.current.onopen = null
    socketRef.current.close()
    socketRef.current = null
  }
}
